c++/src/kj/compat/http.c++

```diff
@@ -5716,14 +5716,111 @@ public:
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->tryRead(buffer, minBytes, maxBytes);
+              } else {
+                return unencryptedStream->tryRead(buffer, minBytes, maxBytes);
+              }
+            }
+
+            kj::Maybe<uint64_t> tryGetLength() override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->tryGetLength();
+              } else {
+                return unencryptedStream->tryGetLength();
+              }
+            }
+
+            kj::Promise<uint64_t> pumpTo(kj::AsyncOutputStream& output, uint64_t amount) override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->pumpTo(output, amount);
+              } else {
+                return unencryptedStream->pumpTo(output, amount);
+              }
+            }
+
+            kj::Promise<void> write(const void* buffer, size_t size) override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->write(buffer, size);
+              } else {
+                return unencryptedStream->write(buffer, size);
+              }
+            }
+
+            kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->write(pieces);
+              } else {
+                return unencryptedStream->write(pieces);
+              }
+            }
+
+            kj::Maybe<kj::Promise<uint64_t>> tryPumpFrom(
+                kj::AsyncInputStream& input, uint64_t amount = kj::maxValue) override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->tryPumpFrom(input, amount);
+              } else {
+                return unencryptedStream->tryPumpFrom(input, amount);
+              }
+            }
+
+            kj::Promise<void> whenWriteDisconnected() override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->whenWriteDisconnected();
+              } else {
+                return unencryptedStream->whenWriteDisconnected();
+              }
+            }
+
+            void shutdownWrite() override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                s->get()->shutdownWrite();
+              } else {
+              unencryptedStream->shutdownWrite();
+              }
+            }
+
+            void abortRead() override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                s->get()->abortRead();
+              } else {
+                unencryptedStream->abortRead();
+              }
+            }
+
+            kj::Maybe<int> getFd() const override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->getFd();
+              } else {
+                return unencryptedStream->getFd();
+              }
+            }
+
+            void startTls(
```

Comment:
Can we at least track whether there is an outstanding read, and throw from this method if it is called while a read is outstanding? Otherwise the outstanding read will end up reading the TLS handshake bytes, breaking everything.

You could have a method like:

```
auto trackRead() {
  KJ_ASSERT(!currentlyReading, "only one read is allowed at a time");
  currentlyReading = true;
  return kj::defer([this]() { currentlyReading = false; });
}
```

Then each read operation would do something like:

```
return inner->read(whatever).attach(trackRead());
```

Note that both `tryRead()` and `pumpTo()` need to be counted here.

Also note `AsyncInputStream` already has a rule that there can only be one read at a time, hence why a boolean is sufficient rather than a counter.